### PR TITLE
server: Add SearchItems implementation

### PR DESCRIPTION
**Pull request by warusadura**
This change add Secret Service method SearchItems and Secret Collection method SearchItems implementation.

### CODE REVIEW

File: server/src/service.rs [hunk 1]
```diff
@@ -83,9 +83,34 @@ impl Service {
...
     ) -> Result<(Vec<OwnedObjectPath>, Vec<OwnedObjectPath>), ServiceError> {
-        todo!()
+        let mut unlocked: Vec<OwnedObjectPath> = Vec::new();
+        let mut locked: Vec<OwnedObjectPath> = Vec::new();
+        let collections = self.collections.lock().await;
+
+        for collection in collections.iter() {
+            for item in collection.inner_items().await {
+                if item.attributes().await == attributes {
+                    if !item.is_locked().await {
```
Comment (bilelmoussaoui) [inline server/src/service.rs:95]: Flip the condition, no need for negation here.

File: server/src/service.rs [hunk 2]
```diff
@@ -83,9 +83,34 @@ impl Service {
     #[zbus(out_args("unlocked", "locked"))]
     pub async fn search_items(
         &self,
-        _attributes: HashMap<&str, &str>,
+        attributes: HashMap<String, String>,
     ) -> Result<(Vec<OwnedObjectPath>, Vec<OwnedObjectPath>), ServiceError> {
-        todo!()
+        let mut unlocked: Vec<OwnedObjectPath> = Vec::new();
```
Comment (bilelmoussaoui) [inline server/src/service.rs:88]: No need to specify the type here

File: server/src/service.rs [hunk 3]
```diff
@@ -83,9 +83,34 @@ impl Service {
...
+        let mut locked: Vec<OwnedObjectPath> = Vec::new();
+        let collections = self.collections.lock().await;
+
+        for collection in collections.iter() {
+            for item in collection.inner_items().await {
+                if item.attributes().await == attributes {
```
Comment (bilelmoussaoui) [inline server/src/service.rs:94]: Is this the correct way to match an item,? I don't think so.

File: server/src/service.rs (hunk 1 above)
Comment (warusadura) [reply server/src/service.rs]: Done, thanks!

File: server/src/service.rs (hunk 2 above)
Comment (warusadura) [reply server/src/service.rs]: Removed, thanks!

File: server/src/collection.rs [hunk 4]
```diff
@@ -46,9 +46,30 @@ impl Collection {
     #[zbus(out_args("results"))]
     pub async fn search_items(
         &self,
-        _attributes: HashMap<String, String>,
+        attributes: HashMap<String, String>,
     ) -> Result<Vec<OwnedObjectPath>, ServiceError> {
-        todo!()
+        let results: Vec<OwnedObjectPath> = self
```
Comment (bilelmoussaoui) [inline server/src/collection.rs:51]:
```suggestion
        let results = self
```

File: server/src/service.rs [hunk 5]
```diff
@@ -83,9 +83,32 @@ impl Service {
...
+                if item.is_locked().await {
+                    locked.push(item.path().clone());
+                } else {
+                    unlocked.push(item.path().clone());
+                }
+            }
+        }
+
+        if unlocked.is_empty() && locked.is_empty() {
+            tracing::info!(
```
Comment (bilelmoussaoui) [inline server/src/service.rs:103]: Both those tracings should be debug imho

Comment (bilelmoussaoui) [reply server/src/service.rs:103]: same for the ones you added in collection.rs

File: server/src/service.rs [hunk 6]
```diff
@@ -83,9 +83,32 @@ impl Service {
...
+        let collections = self.collections.lock().await;
+
+        for collection in collections.iter() {
+            for item in collection.search_inner_items(&attributes).await {
+                if item.is_locked().await {
+                    locked.push(item.path().clone());
+                } else {
+                    unlocked.push(item.path().clone());
+                }
+            }
```
Comment (bilelmoussaoui) [inline server/src/service.rs:99]:
```suggestion

let items = collection.search_inner_items(&attributes).await;
            for item in items {
                if item.is_locked().await {
                    locked.push(item.path().clone());
                } else {
                    unlocked.push(item.path().clone());
                }
            }
```

File: server/src/service.rs (hunk 5 above)
Comment (warusadura) [reply server/src/service.rs]: Done, thanks!

File: server/src/service.rs (hunk 6 above)
Comment (warusadura) [reply server/src/service.rs]: Done, thanks!

**Comment by warusadura** [comment]
todo: https://github.com/bilelmoussaoui/oo7/pull/153#discussion_r1836598146

**Comment by bilelmoussaoui** [comment]
Looks good enough

**Pull request merged**